packages/compass-connections/src/connections-manager.ts

```diff
@@ -0,0 +1,242 @@
+  [ConnectionsManagerEvents.ConnectionAttemptStarted]: (
+    connectionInfoId: ConnectionInfoId
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionAttemptCancelled]: (
+    connectionInfoId: ConnectionInfoId
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionAttemptSuccessful]: (
+    connectionInfoId: ConnectionInfoId,
+    dataService: DataService
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionAttemptFailed]: (
+    connectionInfoId: ConnectionInfoId,
+    error: any
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionDisconnected]: (
+    connectionInfoId: ConnectionInfoId
+  ) => void;
+};
+
+export enum ConnectionStatus {
+  Disconnected = 'disconnected',
+  Connecting = 'connecting',
+  Connected = 'connected',
+  Failed = 'failed',
+}
+
+type ConnectionStatusTransitions = {
+  [event in ConnectionsManagerEvents]?: {
+    [status in ConnectionStatus]?: ConnectionStatus;
+  };
+};
+
+const connectionStatusTransitions: ConnectionStatusTransitions = {
+  [ConnectionsManagerEvents.ConnectionAttemptStarted]: {
+    [ConnectionStatus.Disconnected]: ConnectionStatus.Connecting,
+    [ConnectionStatus.Failed]: ConnectionStatus.Connecting,
+  },
+  [ConnectionsManagerEvents.ConnectionAttemptCancelled]: {
+    [ConnectionStatus.Connecting]: ConnectionStatus.Disconnected,
+  },
+  [ConnectionsManagerEvents.ConnectionAttemptFailed]: {
+    [ConnectionStatus.Connecting]: ConnectionStatus.Failed,
+  },
+  [ConnectionsManagerEvents.ConnectionAttemptSuccessful]: {
+    [ConnectionStatus.Connecting]: ConnectionStatus.Connected,
+  },
+  [ConnectionsManagerEvents.ConnectionDisconnected]: {
+    [ConnectionStatus.Connected]: ConnectionStatus.Disconnected,
+  },
+};
+
+export class ConnectionsManager extends EventEmitter {
+  private connectionAttempts = new Map<ConnectionInfoId, ConnectionAttempt>();
+  private connectionStatuses = new Map<ConnectionInfoId, ConnectionStatus>();
+  private dataServices = new Map<ConnectionInfoId, DataService>();
+
+  constructor(
+    private readonly __TEST_CONNECT_FN?: ConnectFn,
+    private readonly logger?: LoggerAndTelemetry['log']['unbound']
+  ) {
+    super();
+  }
+
+  getDataServiceForConnection(
+    connectionInfoId: ConnectionInfoId
+  ): DataService | undefined {
+    return this.dataServices.get(connectionInfoId);
+  }
+
+  statusOf(connectionInfoId: ConnectionInfoId): ConnectionStatus {
+    return (
+      this.connectionStatuses.get(connectionInfoId) ??
+      ConnectionStatus.Disconnected
+    );
+  }
+
+  cancelAllConnectionAttempts(): void {
+    for (const connectionInfoId of this.connectionAttempts.keys()) {
+      this.cancelConnectionAttempt(connectionInfoId);
+    }
+  }
+
+  /**
+   * @param connectionInfo The adjusted ConnectionInfo object that already has
+   * parameters such as appName.
+   */
+  async connect(connectionInfo: ConnectionInfo): Promise<void> {
+    try {
+      if (
+        [ConnectionStatus.Connecting, ConnectionStatus.Connected].includes(
```

Comment:
So if you call connect function and connection is still in progress, this method just immediately resolves, this doesn't sound right, is this a behavior we really expect? If I awated on connect function and it didn't throw, I'd still need to double-check that I have dataService available?

Reply:
Yea I think I should lean back on my state verification to just throw in this case instead of doing this check because this(connecting to connecting) is an unexpected state.